driver/tensor_driver.hpp

```diff
@@ -125,14 +125,14 @@ int SetTensorNd(miopenTensorDescriptor_t t,
     }
 
     // Dimension lengths vector 'len' comes with a default layout.
-    std::string len_layout = tensor_layout_get_default(layout.size());
+    std::string len_layout = miopen::tensor_layout_get_default(layout.size());
```

Comment:
I guess if change to enum it will go back to the starting point of previous discussion of support multiple layout like here https://github.com/ROCmSoftwarePlatform/MIOpen/pull/343#discussion_r474778988. I mean, currently all implementation is based on stride, and then permute/shuffle the layout string then return the final string interpretation.

Reply:
Agreed with @carlushuang switching to enum everywhere will make us lose the ability to execute arbitrary layout (Unless we want to specify 4 * 3 * 2 *1 = 24 layouts for 2d tensors and even more for 3d tensors.

Reply:
What about format like this:
```C+
enum layout : uint32_t {
    NCHW  = 0x00010203ULL,
    NHWC  = 0x00020301ULL,
    NCDHW = 0x0001020304ULL,
};
```
and so on. It is functionally equivalent to strings. You can shuffle bytes like symbols in chars. Supports up to 8 dims. You can define only required enum members. And comparison is fast.

Reply:
I dont think the strings here are substantially slower because they are fairly small(no heap allocations, and single word comparisons). What might be slower is doing `layout == "NCHW"` since it will convert `NCHW` to `std::string` and will call `strlen` on the string. This could be mitigated by making a string const(ie `const std::string NCHW = "NCHW"`). and doing `layout == NCHW` instead.

Alternatively, we could look into packing these into a 4 byte integer(with 4 bits for each permutation axis), but I don't think that would be faster(and would be a more complicated implementation). We tried to do similar optimizations in the past with network_config string by replacing it with integers. The integer implementation surprisingly wasn't faster than string concat+integer-to-string conversions. So I dont see this as a worthwhile optimization to pursue.

Reply:
I am in doubt too. I am not going to insist unless you think it's worth trying this.

Btw have you noted that enum members hold permutation indices in individual bytes? This makes possible additional optimization, right?

> This could be mitigated by making a string const(ie `const std::string NCHW = "NCHW"`). and doing `layout == NCHW` instead.

Not fully, I guess, because string comparison needs to compare string lengths anyway (or to compare up to the ending `'\0'`, which means reading the 2nd word from memory).

> Alternatively, we could look into packing these into a 4 byte integer(with 4 bits for each permutation axis), but I don't think that would be faster

Yes, we need shifts to access nibbles ;(

Reply:
> What about format like this:

NCDHW overflows the 32-bit integer in your snippet.

> Not fully, I guess, because string comparison needs to compare string lengths anyway (or to compare up to the ending '\0', which means reading the 2nd word from memory).

The lengths are already stored in `std::string` so a call to `strlen` is not necessary(only when constructing from a string literal). Since the sizes are known, the comparison then can be done with a `memcmp` which is pretty fast(and its equivalent to comparing 32-bit integers for 4-byte strings).

Reply:
> NCDHW overflows the 32-bit integer in your snippet.

That is why `uint64_t` should be used (and ULL suffix). Typo fixed))

Reply:
Ok, let's postpone this idea for now.